### PR TITLE
Utility methods to support range selections in editors

### PR DESCRIPTION
**Pull request by CheatEx**
utility method to select ranges of Composites between given bounds; inclusive ancestors

### CODE REVIEW

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java [hunk 1]
```diff
@@ -420,6 +433,82 @@ public void remove() {
     };
   }
 
+  private static <ValueT> Iterable<ValueT> iterateFrom(final ValueT initial, final Function<ValueT, ValueT> trans) {
```
Comment (glebleonov) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: we should get rid of duplication of iterate() method here

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java [hunk 2]
```diff
@@ -420,6 +433,82 @@ public void remove() {
...
+          @Override
+          public void remove() {
+            throw new UnsupportedOperationException();
+          }
+        };
+      }
+    };
+  }
+
+  public static <CompositeT extends NavComposite<CompositeT>>
```
Comment (glebleonov) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: This method has a complex contract for returning list.  We should document it.

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java [hunk 3]
```diff
@@ -420,6 +433,82 @@ public void remove() {
...
+      List<CompositeT> allBetween(CompositeT from, CompositeT to) {
+    List<CompositeT> res = new ArrayList<>();
+    res.add(from);
```
Comment (glebleonov) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: I think it is better to exclude "from" and "to" objects from resulting list

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java [hunk 4]
```diff
@@ -420,6 +433,82 @@ public void remove() {
...
+
+  private static <CompositeT extends NavComposite<CompositeT>>
+      boolean includeClosed(CompositeT left, CompositeT to, List<CompositeT> res) {
+    for (CompositeT next = left.nextSibling(); next != null; next = next.nextSibling()) {
+      if (includeOpen(next, to, res)) {
+        return true;
+      }
+    }
+
+    if (left.getParent() == null) {
```
Comment (glebleonov) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: First, it is better to assert something in the beginning on the method. Second, left may be root in the first call of this method , I think.

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java [hunk 5]
```diff
@@ -420,6 +433,82 @@ public void remove() {
...
+
+    if (to != from) {
+      includeClosed(from, to, res);
+    }
+
+    return res;
+  }
+
+  private static <CompositeT extends NavComposite<CompositeT>>
+      boolean includeClosed(CompositeT left, CompositeT to, List<CompositeT> res) {
```
Comment (glebleonov) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: we can return void here

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java [hunk 6]
```diff
@@ -420,6 +433,82 @@ public void remove() {
...
+
+    if (left.getParent() == null) {
+      throw new IllegalArgumentException("Right bound not found in left's bound hierarchy. to=" + to);
+    }
+
+    return includeClosed(left.getParent(), to, res);
+  }
+
+  private static <CompositeT extends NavComposite<CompositeT>>
+      boolean includeOpen(CompositeT node, CompositeT to, List<CompositeT> res) {
```
Comment (glebleonov) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: I still think, that we can return void

File: model/src/test/java/jetbrains/jetpad/model/composite/CompositesBetweenTest.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+  @Test
+  public void same() {
+    SimpleComposite root = new SimpleComposite("root");
+    assertEquals(asList(root), Composites.allBetween(root, root));
+
+    assertEquals(asList(g), Composites.allBetween(g, g));
+  }
+
+  @Test(expected=IllegalArgumentException.class)
+  public void nonExisting() {
```
Comment (glebleonov) [inline model/src/test/java/jetbrains/jetpad/model/composite/CompositesBetweenTest.java]: may be some more tests for illegal cases will be useful, like allBetween(f, e) or allBetween(b, a)

Comment (CheatEx) [inline model/src/test/java/jetbrains/jetpad/model/composite/CompositesBetweenTest.java]: Added `reversed` test.

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java (hunk 5 above)
Comment (CheatEx) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: Done.

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java (hunk 4 above)
Comment (CheatEx) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: Not sure what should be asserted here. I don't see any problem with root in left.

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java (hunk 3 above)
Comment (CheatEx) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: Done.

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java (hunk 1 above)
Comment (CheatEx) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: Done.

File: model/src/main/java/jetbrains/jetpad/model/composite/Composites.java (hunk 2 above)
Comment (CheatEx) [inline model/src/main/java/jetbrains/jetpad/model/composite/Composites.java]: See the patch.

**Pull request merged**